### PR TITLE
ENH: ShapedImageNeighborhoodRange now supports any buffered region index

### PR DESCRIPTION
**Pull request by N-Dekker**
Dropped the implicit assumption that the buffered region of an image, passed as argument to one of the range constructors, starts at index zero. With this commit, ShapedImageNeighborhoodRange supports any arbitrary buffered region index (which is already supported by "old-school" ITK iterators, like itk::ConstNeighborhoodIterator).

### CODE REVIEW

**Comment by thewtex** [comment]
Sweet! :candy: 

Build details [can be found here](https://open.cdash.org/index.php?project=Insight&filtercount=1&showfilters=1&field1=buildname&compare1=63&value1=ShapedImageNeighborhoodRange-arbitrary-index-BufferedRegion).

**Comment by N-Dekker** [comment]
> Build details [can be found here](https://open.cdash.org/index.php?project=Insight&filtercount=1&showfilters=1&field1=buildname&compare1=63&value1=ShapedImageNeighborhoodRange-arbitrary-index-BufferedRegion).

Thank you, Matt! I'm surprised by the test failure at https://open.cdash.org/testDetails.php?test=712811110&build=5656366 !

File: Modules/Core/Common/test/itkShapedImageNeighborhoodRangeGTest.cxx [hunk 1]
```diff
@@ -744,3 +746,63 @@ TEST(ShapedImageNeighborhoodRange, ConstructorSupportsRValueShapeOffsets)
...
+
+  // Check to see that this range yields the same pixel value as an
+  // itk::ConstNeighborhoodIterator with this arbitrary buffered region index.
+  using ConstNeighborhoodIteratorType =
+    itk::ConstNeighborhoodIterator<ImageType, itk::ConstantBoundaryCondition<ImageType>>;
+  ConstNeighborhoodIteratorType neighborhoodIterator(radius, image, image->GetRequestedRegion());
+
+  range.SetLocation(zeroIndex);
+  neighborhoodIterator.SetLocation(zeroIndex);
+  EXPECT_EQ(range[0], neighborhoodIterator.GetPixel(0));
```
Comment (N-Dekker) [inline Modules/Core/Common/test/itkShapedImageNeighborhoodRangeGTest.cxx]: https://open.cdash.org/testDetails.php?test=712811110&build=5656366 reports a test failure: 

> Expected equality of these values:
>   range[0]
>     Which is: 0
>   neighborhoodIterator.GetPixel(0)
>     Which is: 175366032

The "old-school" `neighborhoodIterator.GetPixel(0)` appears to return an uninitialized value! What did I do wrong?

Comment (phcerdan) [reply Modules/Core/Common/test/itkShapedImageNeighborhoodRangeGTest.cxx]: Didn't read all, but have you initialized the image?
line 770:
```cpp
  const auto image = ImageType::New();
  image->SetRegions(bufferedRegion);
  image->Allocate(true);
  //FillBuffer(0)?
```

Comment (phcerdan) [reply Modules/Core/Common/test/itkShapedImageNeighborhoodRangeGTest.cxx]: Maybe the initialization of the whole image is not needed, but I guess zeroIndex should be initialized.
I guess `neighborhoodIterator.GetPixel(0)` pointer to the center of the neighborhood, ie, `zeroIndex`.

Comment (N-Dekker) [reply Modules/Core/Common/test/itkShapedImageNeighborhoodRangeGTest.cxx]: @phcerdan Can you please have another look at the test that I wrote for this patch: https://github.com/InsightSoftwareConsortium/ITK/blob/f1d3a00f3a0fccd716d5c9ace78f5c3435b2bccf/Modules/Core/Common/test/itkShapedImageNeighborhoodRangeGTest.cxx#L752 

The image type is `itk::Image<int>`, that should have all of its pixels zero-initialized by `image->Allocate(true)`, right?

With a buffered region of size 2 x 3, located at index [10, -1], `zeroIndex` should be outside of the buffered region, right?

So then the pixel value at  `zeroIndex` depends on the "boundary condition", right? In this case, I choose `itk::ConstantBoundaryCondition`, which should yield pixel value zero, by default, right?

I'm just asking because I could be wrong, of course! You know, I can make horrible bugs!

Comment (phcerdan) [reply Modules/Core/Common/test/itkShapedImageNeighborhoodRangeGTest.cxx]: Right! Good analysis. Not sure if that's a bug, or just we are breaking the preconditions of SetLocation.

The `itk::ConstNeighborhoodIterator::m_NeedToUseBoundaryCondition` gets set in the constructor inside the `ConstNeighborhoodIterator::SetRegion` method.
 With the test `ConstNeighborhoodIteratorType neighborhoodIterator(radius, image, image->GetRequestedRegion());`
the flag to use boundary conditions is correctly set to `false`. Try another region in the third argument (bigger, or not overlapping with the image), or maybe even setting it manually will do: `neighborhoodIterator.NeedToUseBoundaryConditionOn()`

Having said that, I am not sure if  `ConstNeighborhoodIterator::SetLocation(p)` has the precondition that the location p has to be inside the image. Try setting the flag manually, or change the region.

Comment (N-Dekker) [reply Modules/Core/Common/test/itkShapedImageNeighborhoodRangeGTest.cxx]: @phcerdan Thanks for your analysis as well  `:-)`. Now I also think the test might be breaking a preconditions of `SetLocation`. (If so, does anyone know if this precondition is documented?) It could be that the function just was not designed to set the location outside of the image. Indeed, the test could be fixed by calling `neighborhoodIterator.NeedToUseBoundaryConditionOn()`, but that feels like cheating.  https://itk.org/ITKSoftwareGuide/html/Book1/ITKSoftwareGuide-Book1ch6.html#x51-1160006.4 says that that method _should be used with caution_... Later this week, I'll try to amend the test so that it would no longer violate this possible precondition.

Comment (phcerdan) [reply Modules/Core/Common/test/itkShapedImageNeighborhoodRangeGTest.cxx]: I think it is just an optimization flag, to not even check the boundaries, based on the regions of the constructor. No cheating setting it to `true`, I think it just enforces extra checkings like `IsInBounds`.

**Comment by N-Dekker** [comment]
Oops, I did something wrong! I just wanted to rebase, and then amend the one commit of this pull request. Now I got two commits in this pull request!

**Comment by N-Dekker** [comment]
@hjmjohnson Do you have a suggestion how to fix this pull request? It has two commits now, but it should have only one. (The one whose subject corresponds to the pull request is the right one. The other one, "BUG: Fixed...parameter" is already at the master.)

**Comment by dzenanz** [comment]
How to squash the last x commits:
https://stackoverflow.com/questions/5189560/squash-my-last-x-commits-together-using-git

**Comment by N-Dekker** [comment]
> 
> 
@dzenanz 
> How to squash the last x commits:
> https://stackoverflow.com/questions/5189560/squash-my-last-x-commits-together-using-git

Thanks, @dzenanz, but I do not want to squash my commits. I only want to have this commit in the pull request: ce97d4ffe4ec46c6e0a535231a59786a87c9c8b3

**Comment by N-Dekker** [comment]
Thanks, @dzenanz!
  
With the amend I just review-pushed, I did:
- a small adjustment to the commit message
- fix (hopefully) the test failure at EXPECT_EQ(range[0], neighborhoodIterator.GetPixel(0)), by calling NeedToUseBoundaryConditionOn(), as discussed with @phcerdan 
- fix (hopefully) ITKCommonKWStyleTest failure "itkShapedImageNeighborhoodRange.h:704: error: Line length exceed 209 (max=200)" (https://open.cdash.org/testDetails.php?test=712836254&build=5656457)
